### ISSUE TITLE
SpaceAfterCStyleCast: false not work with typedef/define in C coding

### ISSUE THREAD

**Issue by Etern213** [llvm/llvm-project#101263]
clang-format options:
```
SpaceAfterCStyleCast: false
```
code:
```
typedef int INT;
#define void VOID;
(int)TestFunc(...)
(INT)TestFunc(...)
(void)TestFunc(...)
(VOID)TestFunc(...)
```
The code above will be formatted as:
```
(int)TestFunc(...)
(INT) TestFunc(...)
(void)TestFunc(...)
(VOID) TestFunc(...)
```
expected:
```
(int)TestFunc(...)
(INT)TestFunc(...)
(void)TestFunc(...)
(VOID)TestFunc(...)
```

**Comment by Etern213** [comment]
anyone can help?

**Comment by owenca** [comment]
Can't reproduce it with 18.1.8 or trunk. Please provide actual code snippet and clang-format version you used.

**Comment by rymiel** [comment]
I can reproduce with every version of clang-format since clang-format version 9.0.1 all the way up to top of main (b7146aed5b99e6f858f6c7243a2a38406bb41d1c)

**Comment by owenca** [comment]
My bad. We still need the context though. When the function call is used in an assignment statement or passed as an argument to another function call, the output is correct:
```
$ clang-format -version
clang-format version 18.1.8
$ cat foo.cpp
i = (INT)TestFunc(i);
f((INT)TestFunc(i));
$ clang-format -style='{SpaceAfterCStyleCast: false}' foo.cpp | diff foo.cpp -
$ clang-format -style='{SpaceAfterCStyleCast: true}' foo.cpp
i = (INT) TestFunc(i);
f((INT) TestFunc(i));
```

**Comment by Etern213** [comment]
> My bad. We still need the context though. When the function call is used in an assignment statement or passed as an argument to another function call, the output is correct:
> 
> ```
> $ clang-format -version
> clang-format version 18.1.8
> $ cat foo.cpp
> i = (INT)TestFunc(i);
> f((INT)TestFunc(i));
> $ clang-format -style='{SpaceAfterCStyleCast: false}' foo.cpp | diff foo.cpp -
> $ clang-format -style='{SpaceAfterCStyleCast: true}' foo.cpp
> i = (INT) TestFunc(i);
> f((INT) TestFunc(i));
> ```

context like
```
#define void VOID
int NumberFunc(int a)
{
    return a;
}

int main()
{
    (VOID)NumberFunc(3);
    return 0;
}
```

**Comment by owenca** [comment]
> ```
> int main()
> {
>     (VOID)NumberFunc(3);       // clang-format doesn't format this line correct
>     return OK;
> }
> ```

Isn't the cast `(VOID)` superfluous?

**Comment by Etern213** [comment]
> > ```
> > int main()
> > {
> >     (VOID)NumberFunc(3);       // clang-format doesn't format this line correct
> >     return OK;
> > }
> > ```
> 
> Isn't the cast `(VOID)` superfluous?

absolutely not.
pclint and other tools will check function return value. `NumberFunc` has int type, but I just dont care its return value just in `main`(maybe in other places), so `(VOID)` here is necessary.

**Comment by owenca** [comment]
You don't need the cast from the [C language standard](https://en.cppreference.com/w/c/language/attributes/nodiscard) point of view. To turn off lint warnings, something like [this](https://stackoverflow.com/a/11889024) might work.

**Comment by owenca** [comment]
This is probably still a bug in that adding `VOID` to `TypeNames` has no effect:
```
$ echo '(VOID)NumberFunc(3);' | clang-format -style='{TypeNames: [VOID]}'
(VOID) NumberFunc(3);
```

**Comment by Etern213** [comment]
> This is probably still a bug in that adding `VOID` to `TypeNames` has no effect:
> ```
> $ echo '(VOID)NumberFunc(3);' | clang-format -style='{TypeNames: [VOID]}'
> (VOID) NumberFunc(3);
> ```

is there a way to modify typename to let clang-format know `(VOID)`? this is just what I wanted   :)

**Comment by owenca** [comment]
> > This is probably still a bug in that adding `VOID` to `TypeNames` has no effect:
> > ```
> > $ echo '(VOID)NumberFunc(3);' | clang-format -style='{TypeNames: [VOID]}'
> > (VOID) NumberFunc(3);
> > ```
> 
> is there a way to modify typename to let clang-format know `(VOID)`? this is just what I wanted :)

Oops! That was 18.1.8. Version 20 already got it to work:
```
$ clang-format -version
clang-format version 20.0.0git (https://github.com/llvm/llvm-project 986bc3d0719af653fecb77e8cfc59f39bec148fd)
$ cat foo.c
void foo() {
  i = (INT)TestFunc(i);
  f((INT)TestFunc(i));

  (VOID)NumberFunc(3);
}
$ clang-format -style='{TypeNames: [INT, VOID]}' foo.c | diff foo.c -
$ 
```

**Comment by Etern213** [comment]
> > > This is probably still a bug in that adding `VOID` to `TypeNames` has no effect:
> > > ```
> > > $ echo '(VOID)NumberFunc(3);' | clang-format -style='{TypeNames: [VOID]}'
> > > (VOID) NumberFunc(3);
> > > ```
> > 
> > is there a way to modify typename to let clang-format know `(VOID)`? this is just what I wanted :)
> 
> Oops! That was 18.1.8. Version 20 already got it to work:
> ```
> $ clang-format -version
> clang-format version 20.0.0git (https://github.com/llvm/llvm-project 986bc3d0719af653fecb77e8cfc59f39bec148fd)
> $ cat foo.c
> void foo() {
>   i = (INT)TestFunc(i);
>   f((INT)TestFunc(i));
> 
>   (VOID)NumberFunc(3);
> }
> $ clang-format -style='{TypeNames: [INT, VOID]}' foo.c | diff foo.c -
> $ 
> ```

wonderful, thanks a lot

**Comment by owenca** [comment]
> Version 20 already got it to work

Fixed by 0baef3b18cdbbdcf9f07c10607407ad0fb541449.
